src/mpid/ch4/netmod/ofi/init_provider.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "ofi_impl.h"
+#include "ofi_init.h"
+
+/* There are two configurations: with or without RUNTIME_CHECKS.
+ *
+ * 1. Without RUNTIME_CHECKS. Global settings are macros defined in ofi_capabilites.h, and
+ *    MPIDI_OFI_global.settings are not used.
+ *    Global settings are used to init hints, which is used to get the first matching provider.
+ *
+ * 2. With RUNTIME_CHECKS. Macros are redirected to fields in MPIDI_OFI_global.settings.
+ *    a. Find providers based on optimal and minimal settings. Global settings are not used at
+ *       this stage and remain uninitialized. The optimal settings are the default set or the set
+ *       that matches MPIR_CVAR_OFI_USE_PROVIDER.
+ *    b. The selected provider is used to init hints. This step is covoluted since we want to
+ *       reuse the init hints routine that is based on global settings.
+ *       b.1. Initialize global.settings with default (or MPIR_CVAR_OFI_USE_PROVIDER).
+ *       b.2. Init hints using global settings -- this step is the same as without RUNTIME_CHECKS
+ *       b.3. Use the hints to get MPIDI_OFI_global.prov_use. This may take a few try, each time
+ *            relaxing attributes such as tx_attr and domain_attr.
+ *
+ *  The hints is used to get the final MPIDI_OFI_global.prov_use, which is used to open the fabric.
+ *  Global settings are finally updated again with the final prov_use. Note that in case of without
+ *  RUNTIME_CHECKS, global settings are mostly macros so the update had little impact. Some settings
+ *  such as "max_buffered_send" are always updated according to fi_info.
+ *
+ * Another way to understand this, there are three ways of control provider selection.
+ * a. Static configuration without RUNTIME_CHECKS, e.g. `--with-device=ch4:ofi:sockets`
+ *    Most global settings are macro constants. We try get the provider with initialized hints once,
+ *    abort if fails.
+ * b. With RUNTIME_CHECKS, use MPIR_CVAR_OFI_USE_PROVIDER. We set the optimal settings based on
+ *    matching set. Thus, the static settings in ofi_capabilities.h can be used to control settings.
+ * c. With RUNTIME_CHECKS, use FI_PROVIDER. We set the optimal settings based on default set.
```

Comment:
Maybe call out that "default set" has a specific meaning here. Reading this makes (b) and (c) sound very similar unless you know that there is a capability set called "default".